zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreter.java

```diff
@@ -227,25 +233,105 @@ public void open() {
 
   @Override
   public void close() {
-    RemoteInterpreterProcess interpreterProcess = getInterpreterProcess();
+    ExecutorService service = Executors.newSingleThreadExecutor();
+    Future future = null;
+    try
+    {
+      future = service.submit(new CloserTask());
+      future.get(10, TimeUnit.SECONDS);
+      return;
+    }
+    catch(InterruptedException | ExecutionException | TimeoutException e)
+    {
+      if(future != null && !future.isDone())
+      {
+        future.cancel(true);
+      }
+      forceKill();
+    }
+  }
 
-    Client client = null;
-    boolean broken = false;
-    try {
-      client = interpreterProcess.getClient();
-      if (client != null) {
-        client.close(noteId, className);
+  private void forceKill()
+  {
+    RemoteInterpreterProcess process = getInterpreterProcess();
+    try
+    {
+      Client client = process.getClient();
+      int pid = client.getPid();
+      Runtime.getRuntime().exec("kill " + pid);
+      sleep(5000);
+      int status = Runtime.getRuntime().exec("kill -0 " + pid).waitFor();
+      if(status != 0)
+      {
+        return;
       }
-    } catch (TException e) {
-      broken = true;
-      throw new InterpreterException(e);
-    } catch (Exception e1) {
-      throw new InterpreterException(e1);
-    } finally {
-      if (client != null) {
-        interpreterProcess.releaseClient(client, broken);
+      else
+      {
+        Runtime.getRuntime().exec("kill -2 " + pid);
+        sleep(5000);
+        status = Runtime.getRuntime().exec("kill -0 " + pid).waitFor();
```

Comment:
It might be clearer to pull this block into a `boolean isProcessUp(int pid)` method.